File: csharp/StockAnalysis/FundamentalAnalysis.cs
namespace StockAnalysis;

/// <summary>
/// Scores a stock based on fundamental financial metrics from Yahoo Finance.
/// </summary>
public class FundamentalAnalysis
{
    private readonly Dictionary<string, object?> _info;
    public Dictionary<string, object> Metrics { get; } = new();
    public Dictionary<string, Signal> Signals { get; } = new();

    public FundamentalAnalysis(Dictionary<string, object?> info) => _info = info;

    public AnalysisResult ComputeAll()
    {
        Valuation();
        Profitability();
        Growth();
        FinancialHealth();
        Dividends();
        AnalystTargets();
        AggregateSignal();
        return new AnalysisResult { Metrics = Metrics, Signals = Signals };
    }

    private void Valuation()
    {
        double? pe = Dbl("trailingPE"), fwdPe = Dbl("forwardPE"), peg = Dbl("pegRatio");
        double? pb = Dbl("priceToBook"), ps = Dbl("priceToSalesTrailing12Months"), evEbitda = Dbl("enterpriseToEbitda");

        Metrics["P/E (TTM)"] = Fmt(pe);
        Metrics["Forward P/E"] = Fmt(fwdPe);
        Metrics["PEG Ratio"] = Fmt(peg);
        Metrics["P/B"] = Fmt(pb);
        Metrics["P/S"] = Fmt(ps);
        Metrics["EV/EBITDA"] = Fmt(evEbitda);

        int score = 0;
        var reasons = new List<string>();
        if (pe is < 20) { score++; reasons.Add($"P/E {pe:F1} < 20"); }
        else if (pe is > 35) { score--; reasons.Add($"P/E {pe:F1} > 35"); }
        if (peg is < 1.5) { score++; reasons.Add($"PEG {peg:F2} < 1.5"); }
        else if (peg is > 2.5) { score--; reasons.Add($"PEG {peg:F2} > 2.5"); }
        if (pb is < 3) { score++; reasons.Add($"P/B {pb:F2} < 3"); }

        Signals["Valuation"] = score > 0 ? new Signal("Bullish", string.Join("; ", reasons))
            : score < 0 ? new Signal("Bearish", string.Join("; ", reasons))
            : new Signal("Neutral", "Fair valuation");
    }

    private void Profitability()
    {
        double? margin = Dbl("profitMargins"), opMargin = Dbl("operatingMargins");
        double? roe = Dbl("returnOnEquity"), roa = Dbl("returnOnAssets");

        Metrics["Profit Margin"] = Pct(margin);
        Metrics["Operating Margin"] = Pct(opMargin);
        Metrics["ROE"] = Pct(roe);
        Metrics["ROA"] = Pct(roa);

        int score = 0;
        var reasons = new List<string>();
        if (margin is > 0.15) { score++; reasons.Add($"Profit margin {margin:P1}"); }
        if (roe is > 0.15) { score++; reasons.Add($"ROE {roe:P1}"); }
        if (roa is > 0.05) { score++; reasons.Add($"ROA {roa:P1}"); }

        Signals["Profitability"] = score >= 2 ? new Signal("Bullish", string.Join("; ", reasons))
            : score == 0 && margin is < 0 ? new Signal("Bearish", "Negative margins")
            : new Signal("Neutral", reasons.Count > 0 ? string.Join("; ", reasons) : "Average profitability");
    }

    private void Growth()
    {
        double? revGrowth = Dbl("revenueGrowth"), earnGrowth = Dbl("earningsGrowth");

        Metrics["Revenue Growth"] = Pct(revGrowth);
        Metrics["Earnings Growth"] = Pct(earnGrowth);

        int score = 0;
        var reasons = new List<string>();
        if (revGrowth is > 0.10) { score++; reasons.Add($"Revenue growth {revGrowth:P1}"); }
        if (earnGrowth is > 0.10) { score++; reasons.Add($"Earnings growth {earnGrowth:P1}"); }

        Signals["Growth"] = score >= 1 ? new Signal("Bullish", string.Join("; ", reasons))
            : revGrowth is < 0 ? new Signal("Bearish", $"Revenue declining {revGrowth:P1}")
            : new Signal("Neutral", "Moderate growth");
    }

    private void FinancialHealth()
    {
        double? de = Dbl("debtToEquity"), current = Dbl("currentRatio"), quick = Dbl("quickRatio");
        double? fcf = Dbl("freeCashflow"), cash = Dbl("totalCash"), debt = Dbl("totalDebt");

        Metrics["Debt/Equity"] = Fmt(de);
        Metrics["Current Ratio"] = Fmt(current);
        Metrics["Quick Ratio"] = Fmt(quick);
        Metrics["Free Cash Flow"] = HumanNumber(fcf);
        Metrics["Total Cash"] = HumanNumber(cash);
        Metrics["Total Debt"] = HumanNumber(debt);

        int score = 0;
        var reasons = new List<string>();
        if (de is < 100) { score++; reasons.Add($"D/E {de:F1} < 100"); }
        else if (de is > 200) { score--; reasons.Add($"D/E {de:F1} > 200"); }
        if (current is > 1.5) { score++; reasons.Add($"Current ratio {current:F2}"); }
        if (fcf is > 0) { score++; reasons.Add("Positive FCF"); }

        Signals["Financial Health"] = score >= 2 ? new Signal("Bullish", string.Join("; ", reasons))
            : score < 0 ? new Signal("Bearish", string.Join("; ", reasons))
            : new Signal("Neutral", reasons.Count > 0 ? string.Join("; ", reasons) : "Average health");
    }

    private void Dividends()
    {
        double? divYield = Dbl("dividendYield"), payout = Dbl("payoutRatio");

        Metrics["Dividend Yield"] = Pct(divYield);
        Metrics["Payout Ratio"] = Pct(payout);

        Signals["Dividends"] = divYield is > 0.02 ? new Signal("Bullish", $"Yield {divYield:P2}")
            : divYield is > 0 ? new Signal("Neutral", $"Yield {divYield:P2}")
            : new Signal("Neutral", "No dividend");
    }

    private void AnalystTargets()
    {
        double? target = Dbl("targetMeanPrice");
        double? currentPrice = Dbl("currentPrice");
        string? rec = _info.GetValueOrDefault("recommendationKey") as string;

        Metrics["Analyst Target Price"] = Fmt(target);
        Metrics["Current Price"] = Fmt(currentPrice);
        Metrics["Analyst Recommendation"] = rec ?? "N/A";

        if (target.HasValue && currentPrice.HasValue && currentPrice > 0)
        {
            double upside = (target.Value - currentPrice.Value) / currentPrice.Value;
            Metrics["Upside to Target"] = $"{upside:P1}";

            Signals["Analyst Target"] = upside > 0.15 ? new Signal("Bullish", $"{upside:P1} upside to ${target:F2}")
                : upside < -0.10 ? new Signal("Bearish", $"{upside:P1} downside to ${target:F2}")
                : new Signal("Neutral", $"{upside:P1} to target ${target:F2}");
        }
    }

    private void AggregateSignal()
    {
        int bullish = Signals.Values.Count(s => s.Direction == "Bullish");
        int bearish = Signals.Values.Count(s => s.Direction == "Bearish");
        int total = Signals.Count;

        Metrics["_fa_bullish"] = bullish;
        Metrics["_fa_bearish"] = bearish;
        Metrics["_fa_neutral"] = total - bullish - bearish;

        Signals["_FA_Overall"] = bullish > bearish + 1
            ? new Signal("Bullish", $"{bullish}/{total} fundamentals bullish")
            : bearish > bullish + 1
                ? new Signal("Bearish", $"{bearish}/{total} fundamentals bearish")
                : new Signal("Neutral", $"Mixed – {bullish}B/{bearish}S/{total - bullish - bearish}N");
    }

    // ── Helpers ──────────────────────────────────────────────────────────

    private double? Dbl(string key)
    {
        if (_info.TryGetValue(key, out var val) && val is double d) return d;
        return null;
    }

    private static object Fmt(double? val) => val.HasValue ? Math.Round(val.Value, 2) : (object)"N/A";
    private static string Pct(double? val) => val.HasValue ? $"{val:P2}" : "N/A";

    private static string HumanNumber(double? val)
    {
        if (!val.HasValue) return "N/A";
        double abs = Math.Abs(val.Value);
        string sign = val < 0 ? "-" : "";
        if (abs >= 1_000_000_000_000) return $"{sign}${abs / 1_000_000_000_000:F2}T";
        if (abs >= 1_000_000_000) return $"{sign}${abs / 1_000_000_000:F2}B";
        if (abs >= 1_000_000) return $"{sign}${abs / 1_000_000:F2}M";
        return $"{sign}${abs:N0}";
    }
}

File: python/fundamental_analysis.py
"""Fundamental Analysis module - retrieves and scores key financial metrics."""

import yfinance as yf


class FundamentalAnalysis:
    """Pull fundamental data from Yahoo Finance and score the stock."""

    def __init__(self, ticker: yf.Ticker, currency_sym: str = "$"):
        self.ticker = ticker
        self.info = ticker.info
        self.currency_sym = currency_sym
        self.metrics: dict = {}
        self.signals: dict = {}

    def compute_all(self) -> dict:
        self._valuation()
        self._profitability()
        self._growth()
        self._financial_health()
        self._dividends()
        self._analyst_targets()
        self._aggregate_signal()
        return {"metrics": self.metrics, "signals": self.signals}

    # -- Valuation ----------------------------------------------------------

    def _valuation(self):
        pe = self.info.get("trailingPE")
        fwd_pe = self.info.get("forwardPE")
        peg = self.info.get("pegRatio")
        pb = self.info.get("priceToBook")
        ps = self.info.get("priceToSalesTrailing12Months")
        ev_ebitda = self.info.get("enterpriseToEbitda")

        self.metrics["P/E (TTM)"] = _fmt(pe)
        self.metrics["Forward P/E"] = _fmt(fwd_pe)
        self.metrics["PEG Ratio"] = _fmt(peg)
        self.metrics["P/B"] = _fmt(pb)
        self.metrics["P/S"] = _fmt(ps)
        self.metrics["EV/EBITDA"] = _fmt(ev_ebitda)

        # Simple scoring
        score = 0
        reasons = []
        if pe and pe < 20:
            score += 1; reasons.append(f"P/E {pe:.1f} < 20")
        elif pe and pe > 35:
            score -= 1; reasons.append(f"P/E {pe:.1f} > 35")

        if peg and peg < 1.5:
            score += 1; reasons.append(f"PEG {peg:.2f} < 1.5")
        elif peg and peg > 2.5:
            score -= 1; reasons.append(f"PEG {peg:.2f} > 2.5")

        if pb and pb < 3:
            score += 1; reasons.append(f"P/B {pb:.2f} < 3")

        if score > 0:
            self.signals["Valuation"] = ("Bullish", "; ".join(reasons))
        elif score < 0:
            self.signals["Valuation"] = ("Bearish", "; ".join(reasons))
        else:
            self.signals["Valuation"] = ("Neutral", "Fair valuation")

    # -- Profitability ------------------------------------------------------

    def _profitability(self):
        margin = self.info.get("profitMargins")
        op_margin = self.info.get("operatingMargins")
        roe = self.info.get("returnOnEquity")
        roa = self.info.get("returnOnAssets")

        self.metrics["Profit Margin"] = _pct(margin)
        self.metrics["Operating Margin"] = _pct(op_margin)
        self.metrics["ROE"] = _pct(roe)
        self.metrics["ROA"] = _pct(roa)

        score = 0
        reasons = []
        if margin and margin > 0.15:
            score += 1; reasons.append(f"Profit margin {margin:.1%}")
        if roe and roe > 0.15:
            score += 1; reasons.append(f"ROE {roe:.1%}")
        if roa and roa > 0.05:
            score += 1; reasons.append(f"ROA {roa:.1%}")

        if score >= 2:
            self.signals["Profitability"] = ("Bullish", "; ".join(reasons))
        elif score == 0 and margin and margin < 0:
            self.signals["Profitability"] = ("Bearish", "Negative margins")
        else:
            self.signals["Profitability"] = ("Neutral", "; ".join(reasons) if reasons else "Average profitability")

    # -- Growth -------------------------------------------------------------

    def _growth(self):
        rev_growth = self.info.get("revenueGrowth")
        earn_growth = self.info.get("earningsGrowth")
        earn_qtr = self.info.get("earningsQuarterlyGrowth")

        self.metrics["Revenue Growth"] = _pct(rev_growth)
        self.metrics["Earnings Growth"] = _pct(earn_growth)
        self.metrics["Earnings Growth (QoQ)"] = _pct(earn_qtr)

        score = 0
        reasons = []
        if rev_growth and rev_growth > 0.10:
            score += 1; reasons.append(f"Revenue growth {rev_growth:.1%}")
        if earn_growth and earn_growth > 0.10:
            score += 1; reasons.append(f"Earnings growth {earn_growth:.1%}")

        if score >= 1:
            self.signals["Growth"] = ("Bullish", "; ".join(reasons))
        elif rev_growth and rev_growth < 0:
            self.signals["Growth"] = ("Bearish", f"Revenue declining {rev_growth:.1%}")
        else:
            self.signals["Growth"] = ("Neutral", "Moderate growth")

    # -- Financial Health ---------------------------------------------------

    def _financial_health(self):
        de = self.info.get("debtToEquity")
        current = self.info.get("currentRatio")
        quick = self.info.get("quickRatio")
        fcf = self.info.get("freeCashflow")
        total_cash = self.info.get("totalCash")
        total_debt = self.info.get("totalDebt")

        self.metrics["Debt/Equity"] = _fmt(de)
        self.metrics["Current Ratio"] = _fmt(current)
        self.metrics["Quick Ratio"] = _fmt(quick)
        self.metrics["Free Cash Flow"] = _human_number(fcf, self.currency_sym)
        self.metrics["Total Cash"] = _human_number(total_cash, self.currency_sym)
        self.metrics["Total Debt"] = _human_number(total_debt, self.currency_sym)

        score = 0
        reasons = []
        if de is not None and de < 100:
            score += 1; reasons.append(f"D/E {de:.1f} < 100")
        elif de is not None and de > 200:
            score -= 1; reasons.append(f"D/E {de:.1f} > 200")

        if current and current > 1.5:
            score += 1; reasons.append(f"Current ratio {current:.2f}")

        if fcf and fcf > 0:
            score += 1; reasons.append("Positive FCF")

        if score >= 2:
            self.signals["Financial Health"] = ("Bullish", "; ".join(reasons))
        elif score < 0:
            self.signals["Financial Health"] = ("Bearish", "; ".join(reasons))
        else:
            self.signals["Financial Health"] = ("Neutral", "; ".join(reasons) if reasons else "Average health")

    # -- Dividends ----------------------------------------------------------

    def _dividends(self):
        div_yield = self.info.get("dividendYield")
        payout = self.info.get("payoutRatio")

        self.metrics["Dividend Yield"] = _pct(div_yield)
        self.metrics["Payout Ratio"] = _pct(payout)

        if div_yield and div_yield > 0.02:
            self.signals["Dividends"] = ("Bullish", f"Yield {div_yield:.2%}")
        elif div_yield and div_yield > 0:
            self.signals["Dividends"] = ("Neutral", f"Yield {div_yield:.2%}")
        else:
            self.signals["Dividends"] = ("Neutral", "No dividend")

    # -- Analyst targets ----------------------------------------------------

    def _analyst_targets(self):
        target = self.info.get("targetMeanPrice")
        current_price = self.info.get("currentPrice") or self.info.get("regularMarketPrice")
        rec = self.info.get("recommendationKey")
        num_analysts = self.info.get("numberOfAnalystOpinions")

        self.metrics["Analyst Target Price"] = _fmt(target)
        self.metrics["Current Price"] = _fmt(current_price)
        self.metrics["Analyst Recommendation"] = rec or "N/A"
        self.metrics["Number of Analysts"] = num_analysts or "N/A"

        if target and current_price:
            upside = (target - current_price) / current_price
            self.metrics["Upside to Target"] = f"{upside:.1%}"
            sym = self.currency_sym
            if upside > 0.15:
                self.signals["Analyst Target"] = ("Bullish", f"{upside:.1%} upside to {sym}{target}")
            elif upside < -0.10:
                self.signals["Analyst Target"] = ("Bearish", f"{upside:.1%} downside to {sym}{target}")
            else:
                self.signals["Analyst Target"] = ("Neutral", f"{upside:.1%} to target {sym}{target}")

    # -- Aggregation --------------------------------------------------------

    def _aggregate_signal(self):
        bullish = sum(1 for s, _ in self.signals.values() if s == "Bullish")
        bearish = sum(1 for s, _ in self.signals.values() if s == "Bearish")
        total = len(self.signals)
        self.metrics["_fa_bullish"] = bullish
        self.metrics["_fa_bearish"] = bearish
        self.metrics["_fa_neutral"] = total - bullish - bearish

        if bullish > bearish + 1:
            self.signals["_FA_Overall"] = ("Bullish", f"{bullish}/{total} fundamentals bullish")
        elif bearish > bullish + 1:
            self.signals["_FA_Overall"] = ("Bearish", f"{bearish}/{total} fundamentals bearish")
        else:
            self.signals["_FA_Overall"] = ("Neutral", f"Mixed – {bullish}B/{bearish}S/{total - bullish - bearish}N")


# -- Helpers ----------------------------------------------------------------

def _fmt(val):
    if val is None:
        return "N/A"
    if isinstance(val, float):
        return round(val, 2)
    return val


def _pct(val):
    if val is None:
        return "N/A"
    return f"{val:.2%}"


def _human_number(val, currency_sym="$"):
    if val is None:
        return "N/A"
    abs_val = abs(val)
    sign = "-" if val < 0 else ""
    s = currency_sym
    if abs_val >= 1_000_000_000_000:
        return f"{sign}{s}{abs_val / 1_000_000_000_000:.2f}T"
    if abs_val >= 1_000_000_000:
        return f"{sign}{s}{abs_val / 1_000_000_000:.2f}B"
    if abs_val >= 1_000_000:
        return f"{sign}{s}{abs_val / 1_000_000:.2f}M"
    return f"{sign}{s}{abs_val:,.0f}"
